doc/man3/EVP_MAC.pod

```diff
@@ -0,0 +1,235 @@
+EVP_MAC_CTX_copy, EVP_MAC_CTX_mac, EVP_MAC_reset, EVP_MAC_size, EVP_MAC_init,
+EVP_MAC_update, EVP_MAC_final, EVP_MAC_ctrl, EVP_MAC_ctrl_str, EVP_MAC_str2ctrl,
+EVP_MAC_hex2ctrl, EVP_MAC_nid, EVP_MAC_name, EVP_add_mac, EVP_get_macbyname,
+EVP_get_macbynid, EVP_get_macbyobj - EVP MAC routines
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ typedef struct evp_mac_st EVP_MAC;
+ typedef struct evp_mac_ctx_st EVP_MAC_CTX;
+
+ EVP_MAC_CTX *EVP_MAC_CTX_new(const EVP_MAC *mac);
+ EVP_MAC_CTX *EVP_MAC_CTX_new_id(int nid);
+ void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx);
+ int EVP_MAC_CTX_copy(EVP_MAC_CTX *dest, EVP_MAC_CTX *src);
+ const EVP_MAC *EVP_MAC_CTX_mac(EVP_MAC_CTX *ctx);
+ int EVP_MAC_reset(EVP_MAC_CTX *ctx);
+ size_t EVP_MAC_size(EVP_MAC_CTX *ctx);
+ int EVP_MAC_init(EVP_MAC_CTX *ctx);
+ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen);
+ int EVP_MAC_final(EVP_MAC_CTX *ctx, unsigned char *out, size_t *poutlen);
+ int EVP_MAC_ctrl(EVP_MAC_CTX *ctx, int cmd, ...);
+ int EVP_MAC_ctrl_str(EVP_MAC_CTX *ctx, const char *type, const char *value);
+ int EVP_MAC_str2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_hex2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_nid(const EVP_MAC *mac);
+ const char *EVP_MAC_name(const EVP_MAC *mac);
+ int EVP_add_mac(const EVP_MAC *mac);
+ const EVP_MAC *EVP_get_macbyname(const char *name);
+ const EVP_MAC *EVP_get_macbynid(int nid);
+ const EVP_MAC *EVP_get_macbyobj(const ASN1_OBJECT *o);
+
+=head1 DESCRIPTION
+
+These types and functions help the application to calculate MACs of
+different types and with different underlying algorithms if there are
+any.
+
+=head2 Types
+
+B<EVP_MAC> is a type that holds the implementation of a MAC.
+
+B<EVP_MAC_CTX> is a context type that holds internal information used
+to compute MACs from the input.
+
+=head2 Context manipulation functions
+
+EVP_MAC_CTX_new() creates a new context for the MAC type C<mac>.
+EVP_MAC_CTX_new_id() creates a new context for the numerical MAC
+identity <nid>.  The created context can then be used with most other
+functions described here.
+
+EVP_MAC_CTX_free() frees the contents of the context as well as the
+context itself.  B<NULL> is a valid parameter, for which this function
+is a no-op.
+
+EVP_MAC_CTX_copy() makes a deep copy of the C<src> context to the
+C<dest> context.  The C<dest> context I<must> have been created before
+calling this function.
+
+EVP_MAC_CTX_mac() returns the B<EVP_MAC> associated with the context
+C<ctx>.
+
+=head2 Computing functions
+
+EVP_MAC_init() sets up the context with information given through
+diverse controls.  Implementations may vary on this, some of them may
+do their initialization in the controls themselves.  However, this
+should be called at the very least before calling EVP_MAC_update() and
+EVP_MAC_final().
+
+EVP_MAC_reset() resets the computation for the given context.  This
+may not be supported by the underlying implementation.
```

Comment:
I think this should get removed.

Reply:
I think it either it should always be supported, or should be removed.

Reply:
You can also wonder what the difference between init and reset is.

Reply:
Supporting it always could be done, actually.
Is it worth it or is it better to simply removed?

I need help to decide

Reply:
Init resets the key, i.e whatever initial calculations the actual MAC implementation does from the given key will be recalculated. The reset (resume in CMAC speak) should start over with the key material already calculated from last init. It's meant to save cycles.

As far as I know, only HMAC and CMAC supports this. For any other algo, reset will probably just do an init from a saved key.

Reply:
If we don't know that the API is needed, don't implement it seems like a reasonable starting rule.

Reply:
Okie, I think I'll remove this.  For HMAC, it simply zeroes the HMAC context, so it's the same as zallocing the HMAC context again, except for the allocation part...

CMAC is undocumented and we use `CMAC_resume` *even less* than `HMAC_CTX_reset`, so it's as good a reason as any to simply turn a blind eye...